### DISCUSSION TITLE
Notifications buffer doesn't handle encrypted events (Pantalaimon not configured properly)

### DISCUSSION THREAD

**Discussion by butter-radish** [alphapapa/ement.el#290] [Q&A] [answered]
### OS/platform

Linux

### Emacs version and provenance

GNU Emacs 29.4, from GNU Guix, installed with Guix

### Emacs command

Default xinit for EXWM that comes with the GNU Guix System, which starts with this process: `/bin/emacs --eval (cond\ ((file-exists-p\ "~/.exwm")\ (load-file\ "~/.exwm"))\ ((not\ (featurep\ (quote\ exwm)))\ (require\ (quote\ exwm))\ (require\ (quote\ exwm-config))\ (exwm-config-default)\ (message\ (concat\ "exwm\ configuration\ not\ found.\ "\ "Falling\ back\ to\ default\ configuration..."))))`

### Emacs frame type

GUI

### Ement package version and provenance

0.15.1, from GNU Guix, installed with Guix

### Actions taken

After connecting to my Matrix ID, I tried to run ‘ement-notifications’.

### Observed results

Error instead of notifications buffer when I try opening it for the first time. Quitting out of the backtrace creates a (mostly) blank notifications buffer (see §Etc.) upon re-executing ‘ement-notifications’. Killing the (blank) notifications buffer recreates the error running ‘ement-notifications’ the next time, and repeat.

### Expected results

A populated notifications buffer.

### Backtrace

```elisp
(search-failed "> ")
  re-search-forward("> ")
  ement-notifications-log-to-buffer(#s(ement-session :user ... :server ... :token "syt_Zm56_hvFgC..." :transaction-id 1057976981 :rooms ... :next-batch "s5137346139_75..." :device-id "8dc1c360ddc167..." :initial-device-display-name "Ement.el: user..." :has-synced-p t :account-data ... :events #<hash-table equal 610/730 0x10e03621>) #s(ement-notification :room-id "!rGTuczgsIZEVa..." :event ... :readp :json-false))
  run-hook-with-args(ement-notifications-log-to-buffer #s(ement-session :user ... :server ... :token "syt_Zm56_hvFgC..." :transaction-id 1057976981 :rooms ... :next-batch "s5137346139_75..." :device-id "8dc1c360ddc167..." :initial-device-display-name "Ement.el: user..." :has-synced-p t :account-data ... :events #<hash-table equal 610/730 0x10e03621>) #s(ement-notification :room-id "!rGTuczgsIZEVa..." :event ... :readp :json-false))
  ement-notifications-callback(#s(ement-session :user ... :server ... :token "syt_Zm56_hvFgC..." :transaction-id 1057976981 :rooms ... :next-batch "s5137346139_75..." :device-id "8dc1c360ddc167..." :initial-device-display-name "Ement.el: user..." :has-synced-p t :account-data ... :events #<hash-table equal 610/730 0x10e03621>) (... ...))
  apply(ement-notifications-callback (... ...))
  #f(compiled-function (&rest args2) #<bytecode -0xade0d8be6fe5cf0>)(((notifications . ...) (next_token . "4553541953")))
  #f(compiled-function () #<bytecode 0x151bb35b08a05010>)()
  plz--respond(#<process plz-request-curl<1>> #<buffer  *plz-request-curl*-26334> "finished\n")
  apply(plz--respond (#<process plz-request-curl<1>> #<buffer  *plz-request-curl*-26334> "finished\n"))
  timer-event-handler([t 26286 40415 411547 nil plz--respond (#<process plz-request-curl<1>> #<buffer  *plz-request-curl*-26334> "finished\n") nil 377000 nil])
```

### Etc.

The blank notifications buffer consists of a single entry like this (redacted the username):

>   [sender:@example:matrix.org type:m.room.encrypted] 

I am using Pantalaimon—but notifications have always worked before, and messaging encrypted rooms still works.

I had mentioned earlier that this was the same error as my previous bug report, #288, but it’s only now—that that has been fixed—I remember that the error with the notifications buffer was different from the ‘magit-section’ one; the notifications buffer errored out like this ever since I upgraded to 0.15.1. 😅

**Comment by alphapapa** [comment, marked as answer]
Given that the message shows as being still encrypted, it would seem to suggest that you haven't configured Pantalaimon properly for that room's keys, so it can't decrypt the message, and Ement does not handle encrypted events, so it fails to format properly.

**Comment by butter-radish** [comment]
I am able to load and send messages to that room though.

**Reply by alphapapa** [reply]
Nevertheless, that event is not being decrypted by Pantalaimon, which probably indicates that you haven't configured the keys necessary for that event.

**Reply by butter-radish** [reply]
You know what, I just got notifications for encrypted messages from another room in ‘ement-notifications’, so I think you’re right that this is Pantalaimon acting funny. Sorry for bothering you, it’s the timing of it—with my upgrade to 0.15.1—that made me think something had broken in Ement. And thenks for the explanation! <3

**Reply by alphapapa** [reply]
It's no bother.  There's probably room for improvement in how Ement handles these things, i.e. the case in which Pantalaimon isn't yet configured correctly for a room or something.  I haven't actually used Pantalaimon myself though (I have to admit I'm a bit tired of typing that name out, haha), so I'm relying on user reports.

Anyway, based on what I've learned, you probably need to use `panctl` or something like that to receive the keys for that room.

**Reply by butter-radish** [reply]
I’m at a loss for what further configuration to do with Pantalaimon. I’ve been using Ement + Pantalaimon for months, and I can both message and get notifications from the same room that Pantalaimon at first wasn’t able to decrypt. 

I guess I’ll try to compile its master branch, because I think the version packaged in Guix is a bit old now.

> I have to admit I'm a bit tired of typing that name out, haha

As a Pantalaimon (and EXWM) user, I’ll let you into a little trick: type “Pan” and then ‘hippie-expand’ (which is ‘M-/’ for me). :3

**Reply by phil-s** [reply]
> type “Pan” and then ‘hippie-expand’

Or failing that, abbrev-mode? :)

**Reply by butter-radish** [reply]
Well, us Pantalaimon users probably type its name more than “pan”, so I don’t see why an abbrev can’t be made from that. ଘ(∩^o^)⊃━☆゜

**Reply by alphapapa** [reply]
> I’m at a loss for what further configuration to do with Pantalaimon. I’ve been using Ement + Pantalaimon for months, and I can both message and get notifications from the same room that Pantalaimon at first wasn’t able to decrypt.

You might try scrolling back in history, maybe you'll find older messages that use a different key, or something like that.  Or, who knows, maybe it was some kind of random or rare race condition bug in Pantalaimon that didn't load the key before trying to decrypt that event.

> I guess I’ll try to compile its master branch, because I think the version packaged in Guix is a bit old now.

Yes, that might help.

> 
> > I have to admit I'm a bit tired of typing that name out, haha
> 
> As a Pantalaimon (and EXWM) user, I’ll let you into a little trick: type “Pan” and then ‘hippie-expand’ (which is ‘M-/’ for me). :3

If I were typing into this box with Emacs, yes, I've used hippie-expand for a long time.  :)

**Comment by butter-radish** [comment]
Ah, but notifications from unencrypted rooms do show up properly…
